ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/RecoveringSidecarRetriever.java

```diff
@@ -49,39 +57,48 @@ public class RecoveringSidecarRetriever implements DataColumnSidecarRetriever {
   private final DataColumnSidecarDbAccessor sidecarDB;
   private final AsyncRunner asyncRunner;
   private final TimeProvider timeProvider;
-  private final Duration recoverInitiationTimeout;
-  private final Duration recoverInitiationCheckInterval;
-  private final int columnCount;
-  private final int recoverColumnCount;
-
-  private final Set<DataColumnSidecarPromiseWithTimestamp> pendingPromises =
-      ConcurrentHashMap.newKeySet();
+  private final Duration recoveryInitiationTimeout;
+  private final Duration recoveryInitiationCheckInterval;
+  private final int numberOfColumns;
+  private final int numberOfColumnsRequiredToReconstruct;
+
+  private final Set<DataColumnSidecarRequestWithTimestamp> pendingRequests =
+      new ConcurrentSkipListSet<>(
+          // prioritise the earliest requests when checking if recovery needs to be initiated
+          Comparator.comparing(DataColumnSidecarRequestWithTimestamp::timestamp)
+              .thenComparing(DataColumnSidecarRequestWithTimestamp::dataColumnSlotAndIdentifier));
   private final Map<UInt64, RecoveryEntry> recoveryBySlot = new ConcurrentHashMap<>();
 
   private Cancellable cancellable;
 
+  private record DataColumnSidecarRequestWithTimestamp(
+      DataColumnSlotAndIdentifier dataColumnSlotAndIdentifier,
+      SafeFuture<DataColumnSidecar> response,
+      UInt64 timestamp) {}
+
   public RecoveringSidecarRetriever(
       final DataColumnSidecarRetriever delegate,
       final KZG kzg,
       final MiscHelpersFulu specHelpers,
       final CanonicalBlockResolver blockResolver,
       final DataColumnSidecarDbAccessor sidecarDB,
       final AsyncRunner asyncRunner,
-      final Duration recoverInitiationTimeout,
-      final Duration recoverInitiationCheckInterval,
+      final Duration recoveryInitiationTimeout,
+      final Duration recoveryInitiationCheckInterval,
       final TimeProvider timeProvider,
-      final int columnCount) {
+      final int numberOfColumns) {
     this.delegate = delegate;
     this.kzg = kzg;
     this.specHelpers = specHelpers;
     this.blockResolver = blockResolver;
     this.sidecarDB = sidecarDB;
     this.asyncRunner = asyncRunner;
-    this.recoverInitiationTimeout = recoverInitiationTimeout;
-    this.recoverInitiationCheckInterval = recoverInitiationCheckInterval;
+    this.recoveryInitiationTimeout = recoveryInitiationTimeout;
+    this.recoveryInitiationCheckInterval = recoveryInitiationCheckInterval;
     this.timeProvider = timeProvider;
-    this.columnCount = columnCount;
-    this.recoverColumnCount = columnCount / 2;
+    this.numberOfColumns = numberOfColumns;
+    // reconstruction of all columns is possible with >= 50% of the column data
+    this.numberOfColumnsRequiredToReconstruct = Math.ceilDiv(numberOfColumns, 2);
```

Comment:
I'm not sure if it's ever possible from crypto side to make it un-even because it's literally 2x cell extension